### PR TITLE
Support opaque images with background chunks

### PR DESCRIPTION
**Pull request by Drvi**
By mistake, I assumed that opaque PNG image cannot have a background chunk, but they can (to provide suggestion for their surroundings), this lead to a bug where we'd add some padding for alpha values even in case there were none.

I believe this fixes https://github.com/JuliaIO/PNGFiles.jl/issues/53.

cc: @t-bltg @IanButterworth

### CODE REVIEW

**Comment by t-bltg** [comment]
Thanks for providing a hotfix that quickly :100:.
It fixes https://github.com/JuliaIO/PNGFiles.jl/issues/53 on my side. I'm running `Plots` tests suite now to see if this fixes https://github.com/JuliaPlots/Plots.jl/issues/4091.

**Comment by t-bltg** [comment]
Test seems to fail with a different message though:
```
julia: malloc.c:3839: _int_malloc: Assertion `chunk_main_arena (bck->bk)' failed.

signal (6): Aborted
```

**Comment by Drvi** [comment]
Thanks! Will continue investigation tomorrow

**Comment by IanButterworth** [comment]
I thought rolling back PNGFiles didn't fix the segfaults?

Maybe I'm missing something

**Comment by t-bltg** [comment]
> I thought rolling back PNGFiles didn't fix the segfaults?

I checked out 0.3.10 a few hours ago and it was failing as well.

Correction of https://github.com/JuliaIO/PNGFiles.jl/pull/54#issuecomment-1030300706:
I've tested more carefully `Plots.jl` (I think the test weren't picking my patched version (dev) of `PNGFiles`). From the look of it, it looks like this PR fixes our test suite.

**Comment by Drvi** [comment]
Since this seems to break a lot of things and this fix helps, I'm going to merge this today.

Not sure how others feel about the warning message when reading an opaque image with a background chunk. My thinking was that if someone is interested in knowing the suggested background color, the warning would at least tell them what it is. But I think in the future we want to provide some utility functions to query this kind of metadata from images and make them separate from loading.

**Comment by t-bltg** [comment]
> I thought rolling back `PNGFiles` didn't fix the segfaults?

An update to this: using `PNGFiles@0.3.12` and running `Plots` test suite **does not segault locally** (not tested in CI though).

> Since this seems to break a lot of things and this fix helps, I'm going to merge this today.

:+1: since it affects not only `Plots.jl`.

**Comment by Drvi** [comment]
@IanButterworth I think I must've made a mistake while evaluating the older version of PNGFiles with Plots. Sorry for the noise everyone.

**Comment by t-bltg** [comment]
Just thinking out loud here, maybe we can improve testing by loading / saving back a bunch of png files from a database, so that regressions like these can be caught in CI.

I just came across this: http://www.schaik.com/pngsuite/ with a relatively small db (http://www.schaik.com/pngsuite/PngSuite-2017jul19.tgz) 66kB, which should be relatively easy to integrate in the CI actions.

**Comment by Drvi** [comment]
We already use pngsuite for testing, but I guess our tests are not as extensive as they should...

File: src/io.jl [hunk 1]
```diff
@@ -182,10 +181,9 @@ function _load(png_ptr, info_ptr; gamma::Union{Nothing,Float64}=nothing, expand_
         buffer_eltype = UInt8
     end
     
-    is_gray = (color_type == PNG_COLOR_TYPE_GRAY_ALPHA) | (color_type == PNG_COLOR_TYPE_GRAY)
-    if _check_background_load(background, is_gray, valid_PLTE, read_as_paletted)
-        is_transparent || throw(ArgumentError("Only transparent images can have a background"))
+    if read_with_background
         background_color = process_background(png_ptr, info_ptr, _adjust_background_bitdepth(background, bit_depth))
+        is_transparent || @warn("Background color for non-transparent image: $(background_color)")
```
Comment (IanButterworth) [inline src/io.jl:186]: Does the user need to be warned about this? If not, perhaps make it a debug?

Comment (IanButterworth) [reply src/io.jl:186]: I just saw your message above. Yeah, I don't think this needs to be generate log noise. Being a debug would at least make it accessible until we expose an api for it

**Comment by t-bltg** [comment]
> We already use pngsuite for testing, but I guess our tests are not as extensive as they should

Damn, I should have read the tests without speaking. OK, I think I'm going to try to enhance those.

**Pull request merged**